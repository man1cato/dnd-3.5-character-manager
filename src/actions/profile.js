import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
