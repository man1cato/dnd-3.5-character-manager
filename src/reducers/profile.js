export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
