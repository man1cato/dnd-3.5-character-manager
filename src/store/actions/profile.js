import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
