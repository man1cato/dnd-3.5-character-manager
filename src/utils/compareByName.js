export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}