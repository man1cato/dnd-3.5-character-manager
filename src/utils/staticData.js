export const schools = [
   'Abjuration','Clairsentience', 'Conjuration', 'Divination', 
   'Enchantment', 'Evocation', 'Illusion', 'Metacreativity', 
   'Necromancy', 'Psychokinesis', 'Psychometabolism', 'Psychoportation', 
   'Telepathy', 'Transmutation', 'Universal'
]

export const alignments = [
   'Lawful Good', 'Neutral Good', 'Chaotic Good', 
   'Lawful Neutral', 'True Neutral', 'Chaotic Neutral', 
   'Lawful Evil', 'Neutral Evil', 'Chaotic Evil'
]