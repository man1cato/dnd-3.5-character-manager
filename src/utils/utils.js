import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])