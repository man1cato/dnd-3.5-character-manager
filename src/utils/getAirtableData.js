import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}