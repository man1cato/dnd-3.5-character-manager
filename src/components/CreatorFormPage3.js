import React, { useEffect, Fragment } from 'react'
import _ from 'lodash'
import { Field, ErrorMessage } from 'formik'
import { rollDice } from '../utils/utils'
import { abilities } from '../utils/staticData'

const CreatorFormPage3 = ({
	values,
	selectedRace,
	handleChange, 
	setFieldValue, 
	validateForm
}) => {
	useEffect(() => {
		validateForm()
	}, [])

	const raceMods = selectedRace.abilityMods

	return (
		<div>
			<h3 className="row--center">Ability Scores</h3>

			<div className="section form-grid--abilities">
				<h5 className="grid__col1 form-grid__header">Ability</h5>
				<h5 className="grid__col2 form-grid__header">Score</h5>
				<h5 className="grid__col3 form-grid__header">Race Mod</h5>
				<h5 className="grid__col4 form-grid__header">Final</h5>

				{_.keys(abilities).map((abbr) => {
					const fieldName = `abilities.${abbr}.score`
					const raceMod = raceMods[abbr]
					return (
						<Fragment key={abbr}>
							<div className="grid__col1">{abilities[abbr]} ({abbr.toUpperCase()})</div>
							<Field 
								className="grid__col2 text-input"
								name={fieldName}
								id={`${abbr}Input`}
								value={values.abilities[abbr].score.toString()}
								onChange={(e) => {
									const score = Number(e.target.value)
									setFieldValue(`abilities.${abbr}.final`, score + raceMod, false)
									handleChange({ 
										target: { 
											name: e.target.name,
											id: e.target.id,
											value: score
										} 
									})
								}}
							/>	
							<div className="grid__col3">{raceMod}</div>
							<div className="grid__col4">{values.abilities[abbr].final}</div>					
							<ErrorMessage className="grid__col-span4 form-group--error" name={fieldName} component="div" />
						</Fragment>
					)
				})}
			</div>

			<div className="row--center">
				<button
					className="button"
					id={`abilitiesRollButton`}
					type='button'
					onClick={() => {
						_.keys(abilities).forEach((abbr) => {
							const score = rollDice(6, 3)
							setFieldValue(`abilities.${abbr}.score`, score)
							setFieldValue(`abilities.${abbr}.final`, score + raceMods[abbr], false)							
						})
					}}
				>
					Roll Scores
				</button>

			</div>
			
		</div>
	)
}

export default CreatorFormPage3