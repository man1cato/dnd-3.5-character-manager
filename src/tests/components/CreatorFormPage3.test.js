import React from 'react'
import { shallow } from 'enzyme'
import _ from 'lodash'
import CreatorFormPage3 from '../../components/CreatorFormPage3'
import { abilities } from '../../utils/staticData'
import { apiData } from '../utils'
import profile from '../fixtures/profile'


let wrapper, props
beforeAll(async () => {
   const api = await apiData()
   props = {
      values: { abilities: _.mapValues(abilities, () => ({ score: '' })) },
      selectedRace: api.races[profile.race]
   }
   wrapper = shallow(<CreatorFormPage3 {...props} />)
})


test('should render CreatorFormPage3 correctly', () => {
	expect(wrapper).toMatchSnapshot()
})