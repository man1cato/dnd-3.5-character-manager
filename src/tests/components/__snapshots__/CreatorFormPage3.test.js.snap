// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CreatorFormPage3 correctly 1`] = `
<div>
  <h3
    className="row--center"
  >
    Ability Scores
  </h3>
  <div
    className="section grid grid--abilities-set"
  >
    <h5
      className="grid__col1"
    >
      Ability
    </h5>
    <h5
      className="grid__col2"
    >
      Score
    </h5>
    <h5
      className="grid__col3"
    >
      Race Mod
    </h5>
    <h5
      className="grid__col4"
    >
      Final
    </h5>
    <h4
      className="grid__col1"
    >
      Strength
       (
      STR
      )
    </h4>
    <FormikConnect(FieldInner)
      className="grid__col2"
      id="strInput"
      name="abilities.str.score"
      onChange={[Function]}
      value=""
    />
    <div
      className="grid__col3"
    >
      -2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="grid__col-span4"
      component="div"
      name="abilities.str.score"
    />
    <h4
      className="grid__col1"
    >
      Dexterity
       (
      DEX
      )
    </h4>
    <FormikConnect(FieldInner)
      className="grid__col2"
      id="dexInput"
      name="abilities.dex.score"
      onChange={[Function]}
      value=""
    />
    <div
      className="grid__col3"
    >
      2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="grid__col-span4"
      component="div"
      name="abilities.dex.score"
    />
    <h4
      className="grid__col1"
    >
      Constitution
       (
      CON
      )
    </h4>
    <FormikConnect(FieldInner)
      className="grid__col2"
      id="conInput"
      name="abilities.con.score"
      onChange={[Function]}
      value=""
    />
    <div
      className="grid__col3"
    >
      -2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="grid__col-span4"
      component="div"
      name="abilities.con.score"
    />
    <h4
      className="grid__col1"
    >
      Intelligence
       (
      INT
      )
    </h4>
    <FormikConnect(FieldInner)
      className="grid__col2"
      id="intInput"
      name="abilities.int.score"
      onChange={[Function]}
      value=""
    />
    <div
      className="grid__col3"
    >
      2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="grid__col-span4"
      component="div"
      name="abilities.int.score"
    />
    <h4
      className="grid__col1"
    >
      Wisdom
       (
      WIS
      )
    </h4>
    <FormikConnect(FieldInner)
      className="grid__col2"
      id="wisInput"
      name="abilities.wis.score"
      onChange={[Function]}
      value=""
    />
    <div
      className="grid__col3"
    >
      0
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="grid__col-span4"
      component="div"
      name="abilities.wis.score"
    />
    <h4
      className="grid__col1"
    >
      Charisma
       (
      CHA
      )
    </h4>
    <FormikConnect(FieldInner)
      className="grid__col2"
      id="chaInput"
      name="abilities.cha.score"
      onChange={[Function]}
      value=""
    />
    <div
      className="grid__col3"
    >
      0
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="grid__col-span4"
      component="div"
      name="abilities.cha.score"
    />
  </div>
  <div
    className="row--center"
  >
    <button
      id="abilitiesRollButton"
      onClick={[Function]}
      type="button"
    >
      Roll Scores
    </button>
  </div>
</div>
`;
